### PR TITLE
add dev container, run formatter, fix workflows

### PR DESCRIPTION
**Pull request by john0isaac**
- Add DevContainer Configuration
- Fix workflows referencing non existent files `docs/.markdownlint.json` and update versions
- Replace deprecated `ruff check` with `ruff check <PATH_OF_FILE>`
- Add precommit configuration 
- recompile the requirements
- Format python files

### CODE REVIEW

**Comment by john0isaac** [comment]
@kjaymiller can you approve this so I merge the other PR? The workflow is failing I fixes it here.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ARG IMAGE="python:3.12"
```
Comment (kjaymiller) [inline .devcontainer/Dockerfile:1]: can we support python3.13

Comment (john0isaac) [reply .devcontainer/Dockerfile:1]: Sure

Comment (john0isaac) [reply .devcontainer/Dockerfile:1]: no we can't it's not available yet in the container image.
https://github.com/devcontainers/images/tree/main/src/python

Can we merge this now so I can merge the other one?

Comment (kjaymiller) [reply .devcontainer/Dockerfile:1]: Last thing... Can you please update the PR description with a walkthrough of the changes and why you made them?

Comment (john0isaac) [reply .devcontainer/Dockerfile:1]: done
can we merge now?

**Review by kjaymiller — APPROVED** [review]

**Pull request merged**